Late escape home link block

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
This is not a security problem. This PR simply moves escaping of all PHP output to be as "late" as possible. This means we avoid escaping variables until they are output in the HTML markup.

This is a WP Core best practice.

## How has this been tested?

* Add Home link block.
* Check all functionality still works "as was" - especially adding class and style attrs.
* Check all tests pass.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [ ] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

getdave committed:
Escape style and css classes

getdave committed:
Escape label

wp_kses doesn't do escaping AFAIK

getdave commented on packages/block-library/src/home-link/index.php:
```diff
@@ -99,12 +99,12 @@ function block_core_home_link_build_li_wrapper_attributes( $context ) {
-	$css_classes     = trim( implode( ' ', $classes ) ) . ' wp-block-navigation-item';
+	$css_classes     = trim( implode( ' ', $classes ) . ' wp-block-navigation-item' );
 
 	$wrapper_attributes = get_block_wrapper_attributes(
```
`get_block_wrapper_attributes` escapes each attribute. I added escaping in here because in the future we might switch to another method which doesn't escape.

ntsekouras replied:
I think that even if someday we would change to something else, that new function would/should also handle the sanitisation. It feels like we are adding boilerplate code.

getdave replied:
Possibly. But still you can't guarantee that. It's effectively 🤞 that the developer remembers to do escaping in the new function.

Escaping as near to the point of output as possible is the most robust option even if it introduces additional overhead.

ntsekouras replied:
>Escaping as near to the point of output as possible is the most robust option 

I agree with that, but I don't think we should add code because it might be needed in the future.

getdave replied:
Looks like I already removed this.

After some checking I've decided that all usages of `get_block_wrapper_attributes` can be considered safe without having to pre-escape the attributes.

ntsekouras commented on packages/block-library/src/home-link/index.php:
```diff
@@ -129,7 +129,7 @@ function render_block_core_home_link( $attributes, $content, $block ) {
 	$aria_current = is_home() || ( is_front_page() && 'page' === get_option( 'show_on_front' ) ) ? ' aria-current="page"' : '';
 
-	$html = '<li ' . $wrapper_attributes . '><a class="wp-block-home-link__content wp-block-navigation-item__content" rel="home"' . $aria_current;
+	$html = '<li ' . $wrapper_attributes . '><a class="wp-block-home-link__content wp-block-navigation-item__content" rel="home"' . esc_attr( $aria_current );
```
`aria_current` is not user input but defined above. Also `esc_attr` is for the attribute values, no? This produces duplicate `"` now.

getdave replied:
> aria_current is not user input but defined above

Understood, but the best practice in WP Core is to escape at the point of _output_ (i.e. [as late as possible](https://developer.wordpress.org/themes/theme-security/data-sanitization-escaping/#:~:text=esc_textarea(%20%24text%20)%3B%20%3F%3E%3C/textarea%3E-,Tip%3A,Output%20escaping%20should%20occur%20as%20late%20as%20possible.,-Top%20%E2%86%91)). This reduces the risk of missing escaping data.

For example, whilst the aria is defined above now, one day that could change and there could be a different method of computing that value. Therefore if we escape when building the HTML we reduce the risk of regressions in future.

> Also esc_attr is for the attribute values, no? This produces duplicate " now.

Sorry my bad. I will have to escape where the value is being built, or change the way the markup is built. Echoing values directly into HTML string without escaping is probably not ideal though...

getdave replied:
I'm going to leave this unescaped and hard coded for now only because the aria is very unlikely to change and it's hardcoded at the point of output.

ntsekouras reviewed (commented):
Thanks for the PR @getdave! 

I'm not 100% sure but isn't `wp_kses` handling the sanitisation of `label` as well? 🤔

getdave commented:
> I'm not 100% sure but isn't `wp_kses` handling the sanitisation of `label` as well?

I'm going to double check this with some Core folks.

**Update**: we can safetly assumed kses does escaping.

getdave committed:
Avoid redundant escaping ontop of kses

getdave committed:
Remove redundant escaping

getdave committed:
Remove escaping of entire attribute.

Only works on values. As this value is unlikely to change we can omit escaping.

getdave commented:
As this PR has been reduced down to a simple spacing change I'm going to close it out. It's not needed. Thank you for your time @ntsekouras - I do appreciate it 🙇
